### PR TITLE
Issue 9497 - ddoc: add support for handling undefined macros

### PR DESCRIPTION
**Pull request by andralex**
Fresh approach replacing https://github.com/D-Programming-Language/dmd/pull/1653

### CODE REVIEW

**Comment by andralex** [comment]
@AndrejMitrovic: so you say

> Use -Dd${RESULTS_DIR}/compilable. ;)

Not sure I get it - where should I use it?

**Comment by AndrejMitrovic** [comment]
In the top of the test files where it's currently missing the {RESULTS_DIR} part.

**Comment by AndrejMitrovic** [comment]
E.g. [here](https://github.com/andralex/dmd/blob/DDOC_UNDEFINED_MACRO/test/compilable/ddoc9497a.d#L3)

File: src/macro.c [hunk 1]
```diff
@@ -376,7 +376,19 @@ void Macro::expand(OutBuffer *buf, size_t start, size_t *pend,
                     continue;
                 }
 
+                bool macroWasUndefined = false;
                 Macro *m = search(name, namelen);
+
+                if (!m)
+                {
+                    static const char undef[] = "DDOC_UNDEFINED_MACRO";
+                    m = search((const utf8_t *)undef, sizeof(undef) - 1);
```
Comment (yebblies) [inline src/macro.c]: Yes.  `undef` will be `const(char)*` and the length info won't be available.

**Comment by andralex** [comment]
Anything holding this?

File: src/macro.c (hunk 1 above)
Comment (yebblies) [inline src/macro.c]: @andralex Please fix this.

**Comment by AndrejMitrovic** [comment]
I think this is good to go!

**Comment by AndrejMitrovic** [comment]
Auto-merge toggled on

**Comment by yebblies** [comment]
Auto-merge toggled off

**Comment by yebblies** [comment]
No it isn't.  The `sizeof` thing still needs to be fixed.

**Comment by AndrejMitrovic** [comment]
Ah I thought he did since he pinged, my bad.

File: src/macro.c (hunk 1 above)
Comment (MartinNowak) [inline src/macro.c]: > @andralex Please fix this.

What needs to be done? It's not obvious.

```c
m = search(const utf8_t *)&undef[0], strlen(&undef[0]));
```

Comment (yebblies) [inline src/macro.c]: Copy any of the similar code in `doc.c` eg https://github.com/andralex/dmd/blob/DDOC_UNDEFINED_MACRO/src/doc.c#L2229

Comment (andralex) [inline src/macro.c]: urgh

**Comment by AndrejMitrovic** [comment]
Auto-merge toggled on

**Pull request merged**